### PR TITLE
Allow user to specify ports via CLI

### LINKED ISSUE #730: Make it easy to use different ports (than 3000 and 3001) during Wasp development
Right now, Wasp has it hardcoded to use 3001 port for api server (backend) and 3000 for web app (frontend). If those ports are already taken, `wasp start` will fail.

This causes problems in different scenarios:
1. User already has something else running on ports 3000 / 3001.
2. User wants to run two Wasp apps at the same time in parallel.

There are two main solution approaches:
1. Allow user to pick different ports manually.
2. Have Wasp dynamically search for open ports and use those.

## Solution 1.1:
After we merge https://github.com/wasp-lang/wasp/pull/683, users will be able to specify ports manually. For example, if they want to have server run on 4040 and web on 8000, they would specify 
```
PORT=8000
REACT_APP_API_URL=http://localhost:4040
```
env vars in .env.client,
and specify
```
PORT=4040
WASP_WEB_CLIENT_URL=http://localhost:8000
```
env vars in .env.server.
The only thing missing with this approach is that it is not very clear in the docs which env vars are available for configuration on frontend. Right now it is any apps that CRA allows configuring, but we don't point that out anywhere. We should therefore look into documenting this better, so it is clear to people how they can do this.

## Solution 1.2:
We could allow even easier way for them to choose these ports, for example by allowing them to specify ports via CLI, smth like `wasp start --web-client-port=8000 --api-server-port=4000`. Python's http-server does that.

## Solution 2.1:
We could, from Wasp CLI code, look for ports, identify open ones, and then make sure that web client and api server use those. This way we choose them dynamically.
I found some initial code for this here:
 - https://stackoverflow.com/questions/39139787/i-want-to-check-whether-or-not-a-certain-port-is-open-haskell

### PR DESCRIPTION
**Pull request by ussgarci**
# Description

This PR  allows the user to specify ports via CLI, smth like `wasp start --web-client-port=8000 --api-server-port=4000`. 

Fixes #730 

## Type of change

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

**Comment by Martinsos** [comment]
@ussgarci how is this one going? I see you made some initial progress.
This one is a bit harder than the rest you tackled so far, so let me know if you need some guidance / help. If you feel it is too hard at the moment, that is also ok, you can take on smth else.

**Comment by ussgarci** [comment]
Hey @Martinsos , thanks for checking in.  This one is going ok so far but I will definitely reach out if I need help!

File: waspc/cli/exe/Main.hs [hunk 1]
```diff
@@ -122,6 +123,37 @@ dbCli args = case args of
   ["studio"] -> runDbCommand studio
   _ -> printDbUsage
 
+startCli :: [String] -> IO ()
+startCli args = case args of
+  [] -> runCommand $ start (StartOptions Nothing Nothing)
+  ["--web-client-port", clientPort, "--api-server-port", apiPort] -> runCommand $ start (StartOptions (Just clientPort) (Just apiPort))
+  ["--web-client-port", clientPort] -> runCommand $ start (StartOptions (Just clientPort) Nothing)
+  ["--api-server-port", apiPort] -> runCommand $ start (StartOptions Nothing (Just apiPort))
```
Comment (Martinsos) [inline waspc/cli/exe/Main.hs:131]: What is tricky here is that you presume order of args. So right now, they can't put api-server-port first, and then web-client-port.

One way to go about this is to just search for them individually in this list.

So you might want to do smth like this:
```hs
let argPairs = zip args (drop 1 args)
let maybeClientPort = snd <$> find (\("--web-client-port", _) -> True) argPairs
let maybeServerPort = snd <$> find (\("--api-server-port", _) -> True) argPairs
case (maybeClientPort, maybeServerPort) of
  (Nothing, Nothing) -> printStartUsageWithArgs args
  _ -> runCommand $ start (StartOptions maybeClientPort maybeServerPort)
```

This is a bit rough code but I think it should give you an idea!

File: waspc/src/Wasp/Common.hs [hunk 2]
```diff
@@ -13,3 +16,12 @@ data DbMigrationsDir
 
 dbMigrationsDirInWaspProjectDir :: Path' (Rel WaspProjectDir) (Dir DbMigrationsDir)
 dbMigrationsDirInWaspProjectDir = [reldir|migrations|]
+
+data StartOptions = StartOptions
```
Comment (Martinsos) [inline waspc/src/Wasp/Common.hs:20]: These feel a bit specific to be put here, I would probably put them somewhere under Cli/Command/Start ... .

File: waspc/src/Wasp/Common.hs [hunk 3]
```diff
@@ -13,3 +16,12 @@ data DbMigrationsDir
...
+  { webClientPort :: Maybe String,
```
Comment (Martinsos) [inline waspc/src/Wasp/Common.hs:21]: Ideally these would already be parsed somewhere earlier and stored here as a number! Probably Int.

File: waspc/src/Wasp/CompileOptions.hs [hunk 4]
```diff
@@ -22,5 +23,7 @@ data CompileOptions = CompileOptions
     -- CLI commands will almost always compile before they execute to ensure the project is up to date.
     -- This filter function allows callers to ignore certain warnings where they do not make sense.
     -- For example, showing a compilation warning to run `db migrate-dev` when you are running that command.
-    generatorWarningsFilter :: [GeneratorWarning] -> [GeneratorWarning]
+    generatorWarningsFilter :: [GeneratorWarning] -> [GeneratorWarning],
+    -- The user can specify ports when running the command `wasp start`from the cli.
+    startOptions :: StartOptions
```
Comment (Martinsos) [inline waspc/src/Wasp/CompileOptions.hs:28]: What would be somewhat better is if we would not call these `startOptions`, but `ports` or smth like that. Maybe even individually, each port on its own, not sure though. But it is certainly not important that they are coming from the `start`, we don't want to carry that info all the way into generator, it is CLI detail that rest of Wasp should not need to know about.

**Review by Martinsos — COMMENTED** [review]
@ussgarci looks as a good direction! Left some comments, but general idea makes sense!

What you described makes sense. Basically we want to set these ports only during development, specifically during `wasp start`. Which we consider to be "development mode", so that is that.

I wonder what is the best way to set them. You proposed to set them via dotEnv files. I guess that makes sense! Tricky thing is that users are providing them, so should we be reading them? What if they provide PORT number both in .env.server, and via CLI? I guess the one in the CLI wins.
So maybe the easiest thing is to just add a port entry to the end of .env(.client | .server) if specified, and that is it. If they have already put their own, ours will just overwrite it I believe since it is below it.

File: waspc/src/Wasp/Common.hs (hunk 2 above)
Comment (ussgarci) [reply waspc/src/Wasp/Common.hs]: Agreed. Done.

File: waspc/src/Wasp/Common.hs (hunk 3 above)
Comment (ussgarci) [reply waspc/src/Wasp/Common.hs]: Yep, done.

File: waspc/src/Wasp/CompileOptions.hs (hunk 4 above)
Comment (ussgarci) [reply waspc/src/Wasp/CompileOptions.hs]: Good call.  Done.

File: waspc/cli/exe/Main.hs (hunk 1 above)
Comment (ussgarci) [reply waspc/cli/exe/Main.hs]: Clever! :sunglasses: I've incorporated this change.  Thanks for the feedback.

**Comment by ussgarci** [comment]
Hey @Martinsos I don't have enough free time to continue with this issue so I'm going to close this PR but I hope in the future I can contribute to Wasp again.  🐝

**Pull request closed without merging**